core/ipc/twin/common/src/main/java/org/opennms/core/ipc/twin/common/AbstractTwinSubscriber.java

```diff
@@ -30,177 +30,83 @@
-        TwinTracker twinTracker = objMap.get(twinUpdate.getKey());
-        if (twinTracker == null) {
-            if (!twinUpdate.isPatch()) {
-                updateSessions(twinUpdate, twinUpdate.getObject());
-            }
-            return;
-        }
-        if (isObjectUpdated(twinTracker, twinUpdate)) {
-            LOG.trace("Received object update with key {}", twinUpdate.getKey());
-            // No need to update if version we are getting is less than what we have with the same session.
-            if (twinTracker.getSessionId().equals(twinUpdate.getSessionId())
-                    && twinTracker.getVersion() > twinUpdate.getVersion()) {
-                return;
-            }
-            // If this is coming completely from new session, reset tracker.
-            if (!twinTracker.getSessionId().equals(twinUpdate.getSessionId())) {
-                if (!twinUpdate.isPatch()) {
-                    updateSessions(twinUpdate, twinUpdate.getObject());
-                }
-                return;
-            }
-            // We can't apply patch when the version jumps more than one version.
-            if (twinUpdate.getVersion() != twinTracker.getVersion() + 1) {
-                sendRpcRequest(new TwinRequest(twinUpdate.getKey(), twinUpdate.getLocation()));
-                return;
-            }
-            byte[] patchedBytes = applyPatch(twinTracker, twinUpdate);
-            // Invoke RPC when we can't apply patch properly.
-            if (patchedBytes != null) {
-                // Update twin object for sessions.
-                updateSessions(twinUpdate, patchedBytes);
-            } else {
-                sendRpcRequest(new TwinRequest(twinUpdate.getKey(), twinUpdate.getLocation()));
-            }
-        }
-    }
-
-    private byte[] applyPatch(TwinTracker twinTracker, TwinUpdate twinUpdate) {
-        if (!twinUpdate.isPatch()) {
-            return twinUpdate.getObject();
-        }
-        try {
-            JsonNode resultingDiff = objectMapper.readTree(twinUpdate.getObject());
-            JsonNode original = objectMapper.readTree(twinTracker.getObj());
-            JsonPatch patch = JsonPatch.fromJson(resultingDiff);
-            JsonNode resultNode = patch.apply(original);
-            return resultNode.toString().getBytes(StandardCharsets.UTF_8);
-        } catch (Exception e) {
-            LOG.error("Not able to apply patch for key {}", twinUpdate.getKey(), e);
-        }
-        return null;
-    }
-
-    private boolean isObjectUpdated(TwinTracker twinTracker, TwinUpdate twinUpdate) {
-        if (twinUpdate.isPatch()) {
-            return true;
-        }
-        return !Arrays.equals(twinTracker.getObj(), twinUpdate.getObject());
-    }
+        // Consume in thread instead of using broker's callback thread.
+        this.executorService.execute(() -> {
+            final var subscription = this.subscriptions.computeIfAbsent(twinUpdate.getKey(), Subscription::new);
```

Comment:
Should a subscription (without a consumer) be created? When the last consumer of a subscription is closed then the subscription gets removed from the map. This indicates that subscriptions without consumers are not stored.

Reply:
The subscription is created to store incoming values before a consumer is registered.

This is a trade-off between leaking memory and early-accepting values. I see more benefit in early-accepting values to spare some requests. Therefore skipping the removal for now.